### PR TITLE
Type housekeeping.

### PR DESCRIPTION
**Pull request by paulp**
Moved ListOfNil somewhere more generally accessible. No
reason the compiler should hoard it for itself. Flitted
to a few locations with constructs like ".head.head" and
".tail.head" looking for code which could be rewritten. Found
some, admittedly not always making use of ListOfNil.

Made overdue moves of ConstantType(Constant(true|false|null))
to vals in Definitions.

### CODE REVIEW

**Comment by paulp** [comment]
Review by @retronym

**Comment by som-snytt** [comment]
Am I the only one thinking it sounds like the basis for a skit with "The Knights Who Say Nil"?

File: src/reflect/scala/reflect/internal/Definitions.scala [hunk 1]
```diff
@@ -544,7 +548,7 @@ trait Definitions extends api.StandardDefinitions {
     lazy val MacroContextClass            = getClassIfDefined("scala.reflect.macros.Context") // defined in scala-reflect.jar, so we need to be careful
          def MacroContextPrefix           = MacroContextClass.map(sym => getMemberMethod(sym, nme.prefix))
          def MacroContextPrefixType       = MacroContextClass.map(sym => getTypeMember(sym, tpnme.PrefixType))
-         def MacroContextUniverse         = MacroContextClass.map(sym => getMemberMethod(sym, nme.universe))
+         lazy val MacroContextUniverse    = MacroContextClass.map(sym => getMemberMethod(sym, nme.universe))
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/Definitions.scala]: Members should not be `vals` as they can acquire new symbols in different runs if, at least, the enclosing symbol switches from one based on class files to one based on source files. This happens in the IDE. See https://issues.scala-lang.org/browse/SI-7678

Comment (paulp) [inline src/reflect/scala/reflect/internal/Definitions.scala]: I was afraid there was some reason like that. No pattern matching, I may as well be writing in java.

Comment (retronym) [inline src/reflect/scala/reflect/internal/Definitions.scala]: You can keep the pretty match at the cost of a local val.

Comment (paulp) [inline src/reflect/scala/reflect/internal/Definitions.scala]: Already paid the piper.

**Comment by retronym** [comment]
LGTM, thanks!

**Comment by retronym** [comment]
LGTM, still, after the squash.

**Comment by paulp** [comment]
Great, I'll take it.

**Pull request merged**